File: client/site_tests/video_YouTubeMseEme/files/video_YouTubeMseEme.js
// Copyright (c) 2013 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.
"use strict";

(function() {
  window.__eventReporter = {};
  window.__testState = {};

  var video_format = 'video/mp4; codecs="avc1.640028"';
  var audio_format = 'audio/mp4; codecs="mp4a.40.2"';

  function createMediaSource() {
    return new WebKitMediaSource();
  }

  function createVideo() {
    return document.createElement('video');
  }

  function setupVideoAndMs(onSourceopen) {
    var temp_video = createVideo();
    var ms = createMediaSource();
    ms.addEventListener('webkitsourceopen', onSourceopen);
    var ms_url = window.URL.createObjectURL(ms);
    temp_video.src = ms_url;
    return {
      'video': temp_video,
      'ms': ms
    };
  }

  window.__testAttach = function() {
    var ms = createMediaSource();
    ms.addEventListener('webkitsourceopen', function() {
      window.__eventReporter['sourceopen'] = true;
    });

    var video = document.getElementById('main_player');
    video.src = window.URL.createObjectURL(ms);
    video.load();
  };

  window.__testAddSourceBuffer = function() {
    var vm = setupVideoAndMs(function() {
      try {
        var return_value = true;
        return_value = return_value && (vm.ms.sourceBuffers.length === 0);
        vm.ms.addSourceBuffer(audio_format);
        return_value = return_value && (vm.ms.sourceBuffers.length === 1);
        vm.ms.addSourceBuffer(video_format);
        return_value = return_value && (vm.ms.sourceBuffers.length === 2);

        window.__testState['addSourceBuffer'] = return_value;
      }
      catch (e) {
        window.__testState['addSourceBuffer'] = false;
      }
    });
  };

  window.__testAddSupportedFormats = function() {
    var formats = [
      audio_format,
      video_format,
    ];

    var vm = setupVideoAndMs(function() {
      for (var i = 0; i < formats.length; ++i) {
        try {
          vm.ms.addSourceBuffer(formats[i]);
        } catch (e) {
          window.__testState['addSupportedFormats'] = false;
          return;
        }
      }
      window.__testState['addSupportedFormats'] = true;
    });
  };

  window.__testAddSourceBufferException = function() {
    var vm = setupVideoAndMs(function() {
      try {
        vm.ms.addSourceBuffer('^^^');
        window.__testState['addSourceBufferException'] = false;
        return;
      }
      catch (e) {
        if (e.code !== DOMException.NOT_SUPPORTED_ERR) {
          window.__testState['addSourceBufferException'] = false;
          return;
        }
      }

      try {
        var temp_media_source = new WebKitMediaSource();
        temp_media_source.addSourceBuffer(audio_format);
        window.__testState['addSourceBufferException'] = false;
        return;
      }
      catch (e) {
        if (e.code !== DOMException.INVALID_STATE_ERR) {
          window.__testState['addSourceBufferException'] = false;
          return;
        }
      }
      window.__testState['addSourceBufferException'] = true;
    });
  };

  window.__testInitialVideoState = function() {
    var temp_video = createVideo();
    var test_result = true;

    test_result = test_result && isNaN(temp_video.duration);
    test_result = test_result && (temp_video.videoWidth === 0);
    test_result = test_result && (temp_video.videoHeight === 0);
    test_result = test_result &&
        (temp_video.readyState === HTMLMediaElement.HAVE_NOTHING);
    test_result = test_result && (temp_video.src === '');
    test_result = test_result && (temp_video.currentSrc === '');

    window.__testState['initialVideoState'] = test_result;
  };

  window.__testInitialMSState = function() {
    var vm = setupVideoAndMs(function() {
      var test_result = true;
      test_result = test_result && isNaN(vm.ms.duration);
      test_result = test_result && vm.ms.readyState === 'open';
      window.__testState['initialMSState'] = test_result;
    });
  };

  window.__testCanPlayClearKey = function() {
    var temp_video = createVideo();
    return temp_video.canPlayType('video/mp4; codecs="avc1.640028"',
                                  'webkit-org.w3.clearkey') === 'probably' &&
        temp_video.canPlayType('audio/mp4; codecs="mp4a.40.2"',
                               'webkit-org.w3.clearkey') === 'probably';
  };

  window.__testCanNotPlayPlayReady = function() {
    var temp_video = createVideo();
    return temp_video.canPlayType('video/mp4; codecs="avc1.640028"',
                                  'com.youtube.playready') !== 'probably' &&
        temp_video.canPlayType('audio/mp4; codecs="mp4a.40.2"',
                               'com.youtube.playready') !== 'probably';
  };
})();

File: client/site_tests/video_YouTubeMseEme/video_YouTubeMseEme.py
# Copyright (c) 2013 The Chromium Authors. All rights reserved.
# Use of this source code is governed by a BSD-style license that can be
# found in the LICENSE file.

import logging, os, time

from autotest_lib.client.bin import test
from autotest_lib.client.common_lib import error
from autotest_lib.client.common_lib.cros import chrome
from autotest_lib.client.cros import httpd


class video_YouTubeMseEme(test.test):
    """The main test class for MSE/EME.

    """


    version = 1

    PLAYER_PAGE = 'http://localhost:8000/video_YouTubeMseEme.html'
    TEST_JS = 'files/video_YouTubeMseEme.js'

    POLLING_TIME = 0.1


    def init(self, chrome, player_page):
        """Initialization function for this test class.

        @param chrome: An Autotest Chrome instance.
        @param player_page: Dummy HTML file to load.

        """
        self._testServer = httpd.HTTPListener(
                8000, docroot=os.path.join(os.path.dirname(__file__), 'files'))
        self._testServer.run()

        self.tab = chrome.browser.tabs[0]
        self.tab.Navigate(player_page)
        self.tab.WaitForDocumentReadyStateToBeComplete()
        self.load_javascript(self.TEST_JS)


    def check_event_happened(self, event, delay_time_sec=2):
        """A wrapper to check if an event in JS has fired.

        @param event: A string to denote the event to check.
        @param delay_time_sec: Time to wait before querying the test (float).
                This is to give the VM some time to schedule the next execution.

        @returns: A boolean indicating if the event has fired.

        """
        start_time = time.time()
        while time.time() - start_time <= delay_time_sec:
            if self.tab.EvaluateJavaScript(
                    'window.__eventReporter["%s"] === true;' % event):
                return True
            time.sleep(self.POLLING_TIME)
        return False


    def load_javascript(self, sub_path):
        """A wrapper to load a JS file into the current tab.

        @param sub_path: The relative path from the current .py file.

        """
        full_path = os.path.join(os.path.dirname(__file__), sub_path)
        with open(full_path) as f:
            js = f.read()
            self.tab.ExecuteJavaScript(js)
            logging.info('Loaded accompanying .js script.')


    def get_test_state(self, test_name, delay_time_sec=2):
        """A wrapper to check the state of a test in the accompanying JS.

        @param test_name: The name of the test that was ran.
        @param delay_time_sec: Time to wait before querying the test (float).
                This is to give the VM some time to schedule the next execution.

        @returns: A boolean value indicating the success of the test.

        """
        start_time = time.time()
        while time.time() - start_time <= delay_time_sec:
            if self.tab.EvaluateJavaScript(
                    'window.__testState["%s"]' % test_name):
                return True
            time.sleep(self.POLLING_TIME)
        return False


    def test_media_source_presence(self):
        """Tests for the existence of the Media Source Extension.

        """
        self.assert_(
                self.tab.EvaluateJavaScript(
                        'window.WebKitMediaSource !== undefined'),
                msg='test_media_source_presence failed.')


    def test_attach_source(self):
        """Tests if attaching a the MediaSource to the video tag is successful.

        """
        self.tab.ExecuteJavaScript('window.__testAttach();')
        self.assert_(
                self.check_event_happened('sourceopen'),
                msg=('test_attach_source failed since "sourceopen" event did '
                     'not fire.'))


    def test_add_source_buffer(self):
        """Tests adding the source buffer to the MediaSource is successful.

        """
        self.tab.ExecuteJavaScript('window.__testAddSourceBuffer();')
        self.assert_(
                self.get_test_state('addSourceBuffer'),
                msg='test_add_source_buffer failed.')


    def test_add_supported_formats(self):
        """Tests adding supported formats to the MediaSource is successful.

        """
        self.tab.ExecuteJavaScript('window.__testAddSupportedFormats();')
        self.assert_(
                self.get_test_state('addSupportedFormats'),
                msg='test_add_supported_formats failed.')


    def test_add_source_buffer_exception(self):
        """Tests adding the source buffer to an uninitialized MediaSource.

        """
        self.tab.ExecuteJavaScript(
                'window.__testAddSourceBufferException();')
        self.assert_(
                self.get_test_state('addSourceBufferException'),
                msg='test_add_source_buffer_exception failed.')


    def test_initial_video_state(self):
        """Tests the initial states of the video object.

        """
        self.tab.ExecuteJavaScript('window.__testInitialVideoState();')
        self.assert_(
                self.get_test_state('initialVideoState'),
                msg='test_initial_video_state failed.')


    def test_initial_media_source_state(self):
        """Tests the initial states of the MediaSource object.

        """
        self.tab.ExecuteJavaScript('window.__testInitialMSState();')
        self.assert_(
                self.get_test_state('initialMSState'),
                msg='test_initial_media_source_state failed.')


    def test_can_play_clear_key(self):
        """Tests if it's possible to play ClearKey content.

        """
        self.assert_(
                self.tab.EvaluateJavaScript(
                        'window.__testCanPlayClearKey();'),
                msg='test_can_play_clear_key failed.')


    def test_can_not_play_play_ready(self):
        """Tests if it's impossible to play PlayReady.

        """
        self.assert_(
                self.tab.EvaluateJavaScript(
                        'window.__testCanNotPlayPlayReady();'),
                msg='test_can_not_play_play_ready failed.')


    def run_once(self, subtest_name):
        with chrome.Chrome() as cr:
            self.init(cr, self.PLAYER_PAGE)

            try:
                # The control file passes in a test name, which is the name of
                #  the test to run, prepended with 'test_'.
                function_to_call = getattr(self, 'test_' + subtest_name)
                function_to_call()
            except AttributeError:
                # Just in case the input test name was mistyped in the control
                #  file.
                raise error.TestFail('No function named: test_' + subtest_name)
